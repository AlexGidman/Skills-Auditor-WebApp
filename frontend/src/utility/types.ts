export const SYSTEM_ROLES = ["StaffUser", "Manager", "Admin"];
export const STAFF_USER = SYSTEM_ROLES[0];
export const MANAGER_SR = SYSTEM_ROLES[1];
export const ADMIN = SYSTEM_ROLES[2];

export const JOB_ROLES = ["Mid-Level Developer", "Senior Developer", "Manager"];
export const MIDLEVEL_DEVELOPER = JOB_ROLES[0];
export const SENIOR_DEVELOPER = JOB_ROLES[1];
export const MANAGER_JR = JOB_ROLES[2];

export const TOKEN_COOKIE = "token";

// Week in seconds
export const TOKEN_LIFE = 60 * 60 * 24 * 7;

export const SKILL_LEVELS = ["None", "Basic", "Intermediate", "High", "Expert"];
export const NONE = SKILL_LEVELS[0];
export const BASIC = SKILL_LEVELS[1];
export const INTERMEDIATE = SKILL_LEVELS[2];
export const HIGH = SKILL_LEVELS[3];
export const EXPERT = SKILL_LEVELS[4];

export type User = {
    firstName: string;
    lastName: string;
    email: string;
    password: string;
};
