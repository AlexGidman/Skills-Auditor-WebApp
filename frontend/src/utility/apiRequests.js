import axios from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders, id) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders, name) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (authHeaders, id, name) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders, id) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (authHeaders, name, category_id) => {
    const data = {
        name,
        category_id,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (authHeaders, id, name, categoryId) => {
    const data = {
        id,
        name,
        category_id: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders, id) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders,
    id,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        id,
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders,
    forename,
    surname,
    email,
    password,
    job_role,
    system_role,
) => {
    const data = {
        forename,
        surname,
        email,
        password,
        job_role,
        system_role,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders, id) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (authHeaders, userId, reportId) => {
    const data = {
        user_id: userId,
        report_id: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders, id) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders, id) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders,
    user_id,
    skill_id,
    skill_level,
    notes,
    expiry_date,
) => {
    const data = {
        user_id,
        skill_id,
        skill_level,
        notes,
        expiry_date,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (authHeaders, id, skill_level, notes) => {
    const data = {
        id,
        skill_level,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email, password) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
