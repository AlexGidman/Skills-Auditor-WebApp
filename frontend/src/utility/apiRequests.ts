import axios, { AxiosRequestHeaders } from "axios";

const categoryEndpoint = "/api/category";
const skillEndpoint = "/api/skill";
const userEndpoint = "/api/user";
const directReportEndpoint = "/api/directreport";
const loginEndpoint = "/api/login";
const userSkillEndpoint = "/api/userskill";

//CATEGORY
export const getAllCategories = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(categoryEndpoint, authHeaders);
};

export const getCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${categoryEndpoint}/${id}`, authHeaders);
};

export const addCategory = async (authHeaders: AxiosRequestHeaders, name: string) => {
    const data = {
        name,
    };
    return axios.post(`${categoryEndpoint}`, data, authHeaders);
};

export const updateCategory = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
) => {
    const data = {
        id,
        name,
    };
    return axios.put(categoryEndpoint, data, authHeaders);
};

export const deleteCategory = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(categoryEndpoint, { ...authHeaders, data });
};

//SKILL
export const getAllSkills = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(skillEndpoint, authHeaders);
};

export const getSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${skillEndpoint}/${id}`, authHeaders);
};

export const addSkill = async (
    authHeaders: AxiosRequestHeaders,
    name: string,
    categoryId: string,
) => {
    const data = {
        name,
        categoryId,
    };
    return axios.post(`${skillEndpoint}`, data, authHeaders);
};

export const updateSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    name: string,
    categoryId: string,
) => {
    const data = {
        id,
        name,
        categoryId: categoryId,
    };
    return axios.put(skillEndpoint, data, authHeaders);
};

export const deleteSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(skillEndpoint, { ...authHeaders, data });
};

//USER
export const getUserDetails = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userEndpoint}/${id}`, authHeaders);
};

export const getUserFromToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${userEndpoint}/token`, authHeaders);
};

export const updateUserDetails = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        id,
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };

    return axios.put(`${userEndpoint}`, data, authHeaders);
};

export const addUser = async (
    authHeaders: AxiosRequestHeaders,
    firstName: string,
    lastName: string,
    email: string,
    password: string,
    jobRole: string, // TODO: change when job role type changes
    systemRole: string, // TODO: change when job system type changes
) => {
    const data = {
        firstName,
        lastName,
        email,
        password,
        jobRole,
        systemRole,
    };
    return axios.post(userEndpoint, data, authHeaders);
};

export const deleteUser = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userEndpoint, { ...authHeaders, data });
};

//DIRECT REPORTS
export const getAllDirectReports = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${directReportEndpoint}/${id}`, authHeaders);
};

export const addDirectReport = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    reportId: string,
) => {
    const data = {
        userId: userId,
        reportId: reportId,
    };
    return axios.post(directReportEndpoint, data, authHeaders);
};

//STAFF SKILLS
export const getAllStaffSkills = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/byuser/${id}`, authHeaders);
};

export const getStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    return axios.get(`${userSkillEndpoint}/${id}`, authHeaders);
};

export const deleteStaffSkill = async (authHeaders: AxiosRequestHeaders, id: string) => {
    const data = { id };
    return axios.delete(userSkillEndpoint, { ...authHeaders, data });
};

export const addUserSkill = async (
    authHeaders: AxiosRequestHeaders,
    userId: string,
    skillId: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
    expiryDate: Date,
) => {
    const data = {
        userId,
        skillId,
        skillLevel,
        notes,
        expiryDate,
    };
    return axios.post(`${userSkillEndpoint}`, data, authHeaders);
};

export const updateStaffSkill = async (
    authHeaders: AxiosRequestHeaders,
    id: string,
    skillLevel: number, // TODO: chnage this when skill level type changes
    notes: string,
) => {
    const data = {
        id,
        skillLevel,
        notes,
    };

    return axios.put(`${userSkillEndpoint}`, data, authHeaders);
};

//LOGIN

export const postLogin = async (email: string, password: string) => {
    const credentials = { email, password };
    return axios.post(`${loginEndpoint}`, credentials);
};

export const validateToken = async (authHeaders: AxiosRequestHeaders) => {
    return axios.get(`${loginEndpoint}/validate`, authHeaders);
};
