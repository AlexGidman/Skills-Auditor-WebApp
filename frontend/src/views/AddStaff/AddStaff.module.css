.button {
    width: 100%;
    max-width: 18rem;
    margin: var(--s2) auto;
}

.form {
    display: flex;
    flex-wrap: wrap;
    margin: 0 10%;
}

.header {
    margin-top: 10%;
}

.item {
    min-width: 14rem;
    margin: 0 3%;
    flex: 44%;
}

@media (max-width: 900px) {
    .form {
        margin: 0;
    }
}
