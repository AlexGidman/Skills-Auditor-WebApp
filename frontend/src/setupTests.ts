// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as apiRequests from "./utility/apiRequests";
import * as reactCookie from "react-cookie";

import { MANAGER_JR, ADMIN, User } from "./utility/types";

jest.mock("uuid", () => {
    return {
        ...jest.requireActual("uuid"),
        v4: jest.fn().mockReturnValue("uuid"),
    };
});

jest.mock("react-cookie", () => {
    return {
        ...jest.requireActual("react-cookie"),
        useCookies: jest.fn(),
    };
});

export const mockApiRequests = {
    getAllCategories: jest.spyOn(apiRequests, "getAllCategories"),
    getCategory: jest.spyOn(apiRequests, "getCategory"),
    updateCategory: jest.spyOn(apiRequests, "updateCategory"),
    getUserDetails: jest.spyOn(apiRequests, "getUserDetails"),
    updateUserDetails: jest.spyOn(apiRequests, "updateUserDetails"),
    getSkill: jest.spyOn(apiRequests, "getSkill"),
    updateSkill: jest.spyOn(apiRequests, "updateSkill"),
    getAllSkills: jest.spyOn(apiRequests, "getAllSkills"),
    addUser: jest.spyOn(apiRequests, "addUser"),
    addDirectReport: jest.spyOn(apiRequests, "addDirectReport"),
    getUserFromToken: jest.spyOn(apiRequests, "getUserFromToken"),
    getAllDirectReports: jest.spyOn(apiRequests, "getAllDirectReports"),
    getAllStaffSkills: jest.spyOn(apiRequests, "getAllStaffSkills"),
    deleteUser: jest.spyOn(apiRequests, "deleteUser"),
    getStaffSkill: jest.spyOn(apiRequests, "getStaffSkill"),
    updateStaffSkill: jest.spyOn(apiRequests, "updateStaffSkill"),
    postLogin: jest.spyOn(apiRequests, "postLogin"),
    addSkill: jest.spyOn(apiRequests, "addSkill"),
    addCategory: jest.spyOn(apiRequests, "addCategory"),
    deleteStaffSkill: jest.spyOn(apiRequests, "deleteStaffSkill"),
    addUserSkill: jest.spyOn(apiRequests, "addUserSkill"),
};

export const mockAdminUser = {
    id: "1",
    firstName: "Admin",
    lastName: "User",
    email: "Admin@email.com",
    jobRole: MANAGER_JR,
    systemRole: ADMIN,
};

export const mockError = { response: { data: { error: new Error("Oops") }, status: 400 } };

beforeAll(() => {
    console.log = jest.fn();
    console.error = jest.fn();
});

beforeEach(() => {
    jest.spyOn(reactCookie, "useCookies").mockImplementation(() => [
        { token: "Bearer token" },
        jest.fn(),
        jest.fn(),
    ]);
    // @ts-ignore
    mockApiRequests.getUserFromToken.mockImplementation(() =>
        Promise.resolve({ data: mockAdminUser }),
    );
});

afterEach(() => {
    jest.resetAllMocks();
});
