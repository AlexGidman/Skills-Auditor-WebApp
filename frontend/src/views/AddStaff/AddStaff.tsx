import React, { useState, useEffect } from "react";
import { useImmer } from "use-immer";

import { TextInput } from "../../components/TextInput/TextInput";
import { Select } from "../../components/Select/Select";
import { Button } from "../../components/Button/Button";

import {
    useAPI,
    useIsAdminOrManager,
    getSelectOptionsFromArray,
    getAssignableSystemRoleOptions,
    isValidUser,
} from "../../utility/helper";
import { addUser, addDirectReport } from "../../utility/apiRequests";
import styles from "./AddStaff.module.css";

import { useNavigate, useOutletContext } from "react-router-dom";
import { MIDLEVEL_DEVELOPER, STAFF_USER, JOB_ROLES, User } from "../../utility/types";

export const AddStaff = () => {
    useIsAdminOrManager();
    const navigate = useNavigate();
    // @ts-ignore TODO fix type for AppOutletContext here
    const [currentUser, setShowToast] = useOutletContext();
    const [userDetails, setUserDetails] = useImmer<User>({
        firstName: "",
        lastName: "",
        email: "",
        password: "",
        jobRole: MIDLEVEL_DEVELOPER,
        systemRole: STAFF_USER,
    });
    const [formChanged, setFormChanged] = useState(false);
    const [formValid, setFormValid] = useState(true);
    const [navigateBack, setNavigateBack] = useState(false);

    const {
        data: createUserData,
        error: createUserError,
        loading: createUserLoading,
        callback: createUser,
    } = useAPI<User>(addUser, Object.values(userDetails), true);

    const {
        error: createDirectReportError,
        loading: createDirectReportLoading,
        callback: createDirectReport,
    } = useAPI(addDirectReport, [currentUser.id, createUserData ? createUserData?.id : ""], true);

    useEffect(() => {
        // if user created successfully create direct report and navigate
        if (createUserData) {
            createDirectReport();
            setNavigateBack(true);
        }
        // if user create fails, just show a toast error
        if (createUserError) {
            setShowToast({ error: "Failed to create user" });
            navigate("/", { replace: true });
        }
    }, [createUserData, createUserError, navigate, setShowToast]);

    useEffect(() => {
        setFormValid(isValidUser(userDetails) && !!userDetails.password);
        setFormChanged(
            !!userDetails.firstName ||
                !!userDetails.lastName ||
                !!userDetails.email ||
                !!userDetails.password ||
                !!userDetails.jobRole ||
                !!userDetails.systemRole,
        );
    }, [userDetails]);

    useEffect(() => {
        if (navigateBack && !createDirectReportLoading) {
            createDirectReportError
                ? setShowToast({ error: "Failed to add user as direct report" })
                : setShowToast({ success: "Successfully added user as direct report" });
            navigate(-1);
        }
    }, [createDirectReportError, createDirectReportLoading, navigate, navigateBack, setShowToast]);

    return (
        <>
            <header>
                <h1>Add Staff</h1>
            </header>
            <section>
                <form className={styles.form}>
                    <TextInput
                        labelText="First name"
                        className={styles.item}
                        value={userDetails.firstName}
                        onChange={(event) => {
                            setUserDetails((draft) => {
                                draft.firstName = event.target.value;
                            });
                        }}
                        required
                    />
                    <TextInput
                        labelText="Last name"
                        className={styles.item}
                        value={userDetails.lastName}
                        onChange={(event) => {
                            setUserDetails((draft) => {
                                draft.lastName = event.target.value;
                            });
                        }}
                        required
                    />
                    <TextInput
                        labelText="Password"
                        className={styles.item}
                        value={userDetails.password}
                        onChange={(event) => {
                            setUserDetails((draft) => {
                                draft.password = event.target.value;
                            });
                        }}
                        type="password"
                    />
                    <Select
                        options={getSelectOptionsFromArray(JOB_ROLES)}
                        labelText="Job role"
                        className={styles.item}
                        value={userDetails.jobRole}
                        onChange={(event) => {
                            setUserDetails((draft) => {
                                draft.jobRole = event.target.value;
                            });
                        }}
                        required
                    />
                    <TextInput
                        labelText="Email"
                        className={styles.item}
                        value={userDetails.email}
                        onChange={(event) => {
                            setUserDetails((draft) => {
                                draft.email = event.target.value;
                            });
                        }}
                        required
                        type="email"
                        pattern="\S+@\S+\.\S+"
                    />
                    <Select
                        options={getAssignableSystemRoleOptions(currentUser.systemRole)}
                        labelText="System role"
                        className={styles.item}
                        value={userDetails.systemRole}
                        onChange={(event) => {
                            setUserDetails((draft) => {
                                draft.systemRole = event.target.value;
                            });
                        }}
                        required
                    />
                    <Button
                        className={styles.button}
                        loading={createUserLoading || createDirectReportLoading}
                        disabled={!formValid || !formChanged}
                        onClick={(e) => {
                            e.preventDefault();
                            createUser();
                        }}
                    >
                        Add Staff
                    </Button>
                </form>
            </section>
        </>
    );
};
