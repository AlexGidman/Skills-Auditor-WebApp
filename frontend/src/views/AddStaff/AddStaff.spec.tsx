import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { mockAdminUser, mockApiRequests, mockApiResponse, mockError } from "../../setupTests";
import * as ReactDom from "react-router-dom";
import { AddStaff } from "./AddStaff";
import { MIDLEVEL_DEVELOPER, STAFF_USER } from "../../utility/types";
import { wait } from "@testing-library/user-event/dist/utils";
import React from "react";

jest.mock("react-router-dom", () => {
    return {
        ...jest.requireActual("react-router-dom"),
        useOutletContext: jest.fn(),
        useNavigate: jest.fn(),
    };
});

describe("AddStaff", () => {
    beforeEach(() => {
        const mockUseOutletContext = ReactDom.useOutletContext as jest.Mock<any, any>;
        mockUseOutletContext.mockReturnValue({
            currentUser: mockAdminUser,
            setShowToast: jest.fn(),
        });
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => jest.fn());
    });
    it("renders correctly", async () => {
        const { container } = render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();

        expect(screen.getByRole("textbox", { name: "First name:" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "First name:" })).toHaveValue("");
        expect(screen.getByRole("textbox", { name: "Last name:" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "Last name:" })).toHaveValue("");
        expect(screen.getByRole("textbox", { name: "Email:" })).toBeInTheDocument();
        expect(screen.getByRole("textbox", { name: "Email:" })).toHaveValue("");
        expect(screen.getByLabelText(/Password/)).toBeInTheDocument(); // no accessible role for passoword inputs

        expect(screen.getByRole("combobox", { name: "Job role" })).toBeInTheDocument();
        expect(screen.getByRole("combobox", { name: "Job role" })).toHaveValue(MIDLEVEL_DEVELOPER);
        expect(screen.getByRole("combobox", { name: "System role" })).toBeInTheDocument();
        expect(screen.getByRole("combobox", { name: "System role" })).toHaveValue(STAFF_USER);

        expect(screen.getByRole("button", { name: "Add Staff" })).toBeInTheDocument();
        expect(screen.getByRole("button", { name: "Add Staff" })).toBeDisabled();

        expect(container).toMatchSnapshot();
    });

    it("should have Add Staff button that becomes enabled when form is valid and has changed", async () => {
        render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();
        const inputChangeArray = [
            {
                change: () =>
                    userEvent.type(screen.getByRole("textbox", { name: "First name:" }), "a"),
                undo: () => userEvent.clear(screen.getByRole("textbox", { name: "First name:" })),
            },
            {
                change: () =>
                    userEvent.type(screen.getByRole("textbox", { name: "Last name:" }), "a"),
                undo: () => userEvent.clear(screen.getByRole("textbox", { name: "Last name:" })),
            },
            {
                change: () =>
                    userEvent.type(screen.getByRole("textbox", { name: "Email:" }), "a@email.com"),
                undo: () => userEvent.clear(screen.getByRole("textbox", { name: "Email:" })),
            },
            {
                change: () => userEvent.type(screen.getByLabelText(/Password/), "aaaaaaaaaa"), // min 10 chars
                undo: () => userEvent.clear(screen.getByLabelText(/Password/)),
            },
        ];
        expect(screen.getByRole("button", { name: "Add Staff" })).toBeDisabled();
        inputChangeArray.forEach((changeInputFunc) => changeInputFunc.change());
        expect(screen.getByRole("button", { name: "Add Staff" })).not.toBeDisabled();
        inputChangeArray.forEach((changeInputFunc) => changeInputFunc.undo());
        expect(screen.getByRole("button", { name: "Add Staff" })).toBeDisabled();
    });

    it("should pass user params to addUser API when Add Staff button clicked", async () => {
        mockApiRequests.addUser.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: { id: 5, message: "Success" } }),
        );
        render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();

        expect(mockApiRequests.addUser).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "First name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Last name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Email:" }), "a@email.com");
        userEvent.type(screen.getByLabelText(/Password/), "aaaaaaaaaa"); // min 10 chars
        userEvent.click(screen.getByRole("button", { name: "Add Staff" }));
        expect(mockApiRequests.addUser).toHaveBeenCalled();
        expect(mockApiRequests.addUser).toHaveBeenLastCalledWith(
            { headers: { authorization: "Bearer token" } },
            "a",
            "a",
            "a@email.com",
            "aaaaaaaaaa",
            MIDLEVEL_DEVELOPER,
            STAFF_USER,
        );
    });

    it("should call addDirectReport API when Add Staff button clicked following addUser called and successfully returns data", async () => {
        const mockNewUserId = 5;

        mockApiRequests.addUser.mockImplementation(() =>
            Promise.resolve({
                ...mockApiResponse,
                data: { id: mockNewUserId, message: "Success" },
            }),
        );

        mockApiRequests.addDirectReport.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: "Success" }),
        );
        render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();

        expect(mockApiRequests.addUser).not.toHaveBeenCalled();
        expect(mockApiRequests.addDirectReport).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "First name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Last name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Email:" }), "a@email.com");
        userEvent.type(screen.getByLabelText(/Password/), "aaaaaaaaaa"); // min 10 chars
        userEvent.click(screen.getByRole("button", { name: "Add Staff" }));
        expect(mockApiRequests.addUser).toHaveBeenCalled();
        await wait(100);
        expect(mockApiRequests.addDirectReport).toHaveBeenCalled();
        expect(mockApiRequests.addDirectReport).toHaveBeenLastCalledWith(
            { headers: { authorization: "Bearer token" } },
            mockAdminUser.id,
            mockNewUserId,
        );
    });

    it("should not call addDirectReport API when Add Staff button clicked following failed addUser call", async () => {
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);
        mockApiRequests.addUser.mockImplementation(() => Promise.reject(mockError));
        render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();

        expect(mockNavigate).not.toHaveBeenCalled();
        expect(mockApiRequests.addUser).not.toHaveBeenCalled();
        expect(mockApiRequests.addDirectReport).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "First name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Last name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Email:" }), "a@email.com");
        userEvent.type(screen.getByLabelText(/Password/), "aaaaaaaaaa"); // min 10 chars
        userEvent.click(screen.getByRole("button", { name: "Add Staff" }));
        expect(mockApiRequests.addUser).toHaveBeenCalled();
        await wait(100);
        expect(mockApiRequests.addDirectReport).not.toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith("/", { replace: true }); // refreshes page if addUser fails
    });

    it("should redirect to previous page when Add Staff button clicked and API call successful", async () => {
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);

        mockApiRequests.addUser.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: { id: 5, message: "Success" } }),
        );

        mockApiRequests.addDirectReport.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: "Success" }),
        );
        render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();
        expect(mockNavigate).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "First name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Last name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Email:" }), "a@email.com");
        userEvent.type(screen.getByLabelText(/Password/), "aaaaaaaaaa"); // min 10 chars
        userEvent.click(screen.getByRole("button", { name: "Add Staff" }));
        await wait(100);
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith(-1);
    });

    it("should redirect to previous page when Add Staff button clicked and API ca;; fails", async () => {
        const mockNavigate = jest.fn();
        jest.spyOn(ReactDom, "useNavigate").mockImplementation(() => mockNavigate);

        mockApiRequests.addUser.mockImplementation(() =>
            Promise.resolve({ ...mockApiResponse, data: { id: 5, message: "Success" } }),
        );
        mockApiRequests.addDirectReport.mockImplementation(() => Promise.reject(mockError));
        render(
            <ReactDom.MemoryRouter>
                <AddStaff />
            </ReactDom.MemoryRouter>,
        );
        expect(await screen.findByRole("heading", { name: "Add Staff" })).toBeInTheDocument();
        expect(mockNavigate).not.toHaveBeenCalled();
        userEvent.type(screen.getByRole("textbox", { name: "First name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Last name:" }), "a");
        userEvent.type(screen.getByRole("textbox", { name: "Email:" }), "a@email.com");
        userEvent.type(screen.getByLabelText(/Password/), "aaaaaaaaaa"); // min 10 chars
        userEvent.click(screen.getByRole("button", { name: "Add Staff" }));
        await wait(100);
        expect(mockNavigate).toHaveBeenCalled();
        expect(mockNavigate).toHaveBeenLastCalledWith(-1);
    });
});
