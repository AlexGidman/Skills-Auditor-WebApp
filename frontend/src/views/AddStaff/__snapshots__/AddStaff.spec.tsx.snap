// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddStaff renders correctly 1`] = `
<div>
  <header>
    <h1>
      Add Staff
    </h1>
  </header>
  <section>
    <form
      class="form"
    >
      <div
        class="component item"
      >
        <label
          class="label"
          for="First-name-undefined"
        >
          First name
          :
        </label>
        <input
          class="input"
          id="First-name-undefined"
          maxlength="100"
          minlength="0"
          pattern="[a-zA-ZÀ-ȕ-]+$"
          required=""
          value=""
        />
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Last-name-undefined"
        >
          Last name
          :
        </label>
        <input
          class="input"
          id="Last-name-undefined"
          maxlength="100"
          minlength="0"
          pattern="[a-zA-ZÀ-ȕ-]+$"
          required=""
          value=""
        />
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Password-undefined"
        >
          Password
          :
        </label>
        <input
          class="input"
          id="Password-undefined"
          maxlength="100"
          minlength="0"
          pattern="[a-zA-ZÀ-ȕ-]+$"
          type="password"
          value=""
        />
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Job-role-undefined"
        >
          Job role
          :
        </label>
        <select
          aria-label="Job role"
          class="select"
          name="Job role"
          required=""
        >
          <option
            value="Mid-Level Developer"
          >
            Mid-Level Developer
          </option>
          <option
            value="Senior Developer"
          >
            Senior Developer
          </option>
          <option
            value="Manager"
          >
            Manager
          </option>
        </select>
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="Email-undefined"
        >
          Email
          :
        </label>
        <input
          class="input"
          id="Email-undefined"
          maxlength="100"
          minlength="0"
          pattern="\\\\S+@\\\\S+\\\\.\\\\S+"
          required=""
          type="email"
          value=""
        />
      </div>
      <div
        class="component item"
      >
        <label
          class="label"
          for="System-role-undefined"
        >
          System role
          :
        </label>
        <select
          aria-label="System role"
          class="select"
          name="System role"
          required=""
        >
          <option
            value="StaffUser"
          >
            StaffUser
          </option>
          <option
            value="Manager"
          >
            Manager
          </option>
          <option
            value="Admin"
          >
            Admin
          </option>
        </select>
      </div>
      <button
        class="component button"
        disabled=""
      >
        Add Staff
      </button>
    </form>
  </section>
</div>
`;
